### PR TITLE
add hybrid parallelism example

### PR DESCRIPTION
**Pull request by sandyhouse**
add the example of using hybrid parallelism training of the ernie model.

recommend config for 4D:
```python
dist_strategy.sharding = True
dist_strategy.sharding_configs = {"segment_broadcast_MB": 32,
"sharding_degree":args.num_sharding,
"mp_degree":args.num_mp
"pp_degree": args.num_pp,
"dp_degree": args.num_dp,
"optimize_offload": True,
}
dist_strategy.pipeline = True
dist_strategy.pipeline_configs = {"schedule_mode": "1F1B",
"micro_batch_size": micro_bsz,
"accumulate_steps": acc_steps,
}
```

recommend config for 3D:
```python
dist_strategy.sharding = True
dist_strategy.sharding_configs = {"segment_broadcast_MB": 32,
"sharding_degree":args.num_sharding,
"mp_degree":args.num_mp
"pp_degree": args.num_pp,
"dp_degree": 1,
"optimize_offload": True,
}
dist_strategy.pipeline = True
dist_strategy.pipeline_configs = {"schedule_mode": "1F1B",
"micro_batch_size": micro_bsz,
"accumulate_steps": acc_steps,
}
```

### CODE REVIEW

File: examples/hybrid_parallelism/run_pretraining.py [hunk 1]
```diff
@@ -0,0 +1,504 @@
...
+                                           "hybrid_dp": False,
+                                           "pp_allreduce_in_optimize": False,
+                                           "gradient_merge_acc_step": 1,
+                                           "optimize_offload": False,
+                                           }
+        dist_strategy.pipeline_configs = {"schedule_mode": "1F1B",
+                                          "micro_batch_size": micro_bsz,
+                                          "accumulate_steps": acc_steps,
+                                          }
+    else:
```
Comment (JZ-LIANG) [inline examples/hybrid_parallelism/run_pretraining.py]: here should be removed to simplify the parallelism case

Comment (JZ-LIANG) [reply examples/hybrid_parallelism/run_pretraining.py]: from line 279 to line 309

File: examples/hybrid_parallelism/run_pretraining.py [hunk 2]
```diff
@@ -0,0 +1,504 @@
...
+    acc_steps = bsz_per_dp // micro_bsz
+
+    # sharding \ model parallel \ pipeline
+    if args.use_sharding:
+        dist_strategy.sharding_configs = {"segment_broadcast_MB": 32,
+                                           "sharding_degree": args.num_sharding,
+                                           "mp_degree": args.num_mp,
+                                           "pp_degree": args.num_pp,
+                                           "dp_degree":args.num_dp,
+                                           "hybrid_dp": False,
```
Comment (JZ-LIANG) [inline examples/hybrid_parallelism/run_pretraining.py]: hybrid_dp setting is deprecate and should be remove

File: examples/hybrid_parallelism/run_pretraining.py [hunk 3]
```diff
@@ -0,0 +1,504 @@
...
+                                           "pp_allreduce_in_optimize": False,
+                                           "gradient_merge_acc_step": 1,
+                                           "optimize_offload": False,
```
Comment (JZ-LIANG) [inline examples/hybrid_parallelism/run_pretraining.py]: optimize_offload is recommended  be true when use pipeline

Comment (JZ-LIANG) [reply examples/hybrid_parallelism/run_pretraining.py]: remove these two setting[gradient_merge_acc_step, pp_allreduce_in_optimize] if we just want to show use case for 3D or  mp-pp

Comment (JZ-LIANG) [reply examples/hybrid_parallelism/run_pretraining.py]: recommend config for 4D: 
       dist_strategy.sharding = True                                     
        dist_strategy.sharding_configs = {"segment_broadcast_MB": 32,
                                            "sharding_degree":args.num_sharding,
                                            "mp_degree":args.num_mp
                                            "pp_degree": args.num_pp,
                                            "dp_degree": args.num_dp,
                                            "optimize_offload": True,
                                            }
        dist_strategy.pipeline = True
        dist_strategy.pipeline_configs = {"schedule_mode": "1F1B",
                                            "micro_batch_size": micro_bsz,
                                            "accumulate_steps": acc_steps,
                                            }


recommend config for 3D: 
       dist_strategy.sharding = True                                     
        dist_strategy.sharding_configs = {"segment_broadcast_MB": 32,
                                            "sharding_degree":args.num_sharding,
                                            "mp_degree":args.num_mp
                                            "pp_degree": args.num_pp,
                                            "dp_degree": 1,
                                            "optimize_offload": True,
                                            }
        dist_strategy.pipeline = True
        dist_strategy.pipeline_configs = {"schedule_mode": "1F1B",
                                            "micro_batch_size": micro_bsz,
                                            "accumulate_steps": acc_steps,
                                            }

File: examples/hybrid_parallelism/run_pretraining.py (hunk 1 above)
Comment (sandyhouse) [reply examples/hybrid_parallelism/run_pretraining.py]: Done.

File: examples/hybrid_parallelism/run_pretraining.py (hunk 2 above)
Comment (sandyhouse) [reply examples/hybrid_parallelism/run_pretraining.py]: Done.

File: examples/hybrid_parallelism/run_pretraining.py (hunk 3 above)
Comment (sandyhouse) [reply examples/hybrid_parallelism/run_pretraining.py]: Done.

File: examples/hybrid_parallelism/pretrain.sh [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+	--log_steps 1 \
+	--num_train_steps 1000000 \
+	--save_steps 100000 \
+	--output_dir ${output_dir} \
+	--use_recompute true \
+	--use_sharding true \
+	--use_sop false \
+	--num_mp=4 \
+        --num_sharding=2 \
+	--num_pp=32 \
```
Comment (JZ-LIANG) [inline examples/hybrid_parallelism/pretrain.sh]: shall we use ernie base  as a simple example here? so as that the example could be run on a 2 nodes cluster with 8 gpus per  node, which is the minimal gpus required to run 4D parallelism

File: examples/hybrid_parallelism/pretrain.sh [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+output_dir=output/${task_name}
+rm -rf ${output_dir}
+
+python -m paddle.distributed.fleet.launch \
+	--log_dir ${output_dir}/log \
+run_pretraining.py \
+	--global_bsz 4096 \
+	--micro_bsz 8 \
+	--max_seq_len 512 \
+	--ernie_config_file config/ernie_230b_config_proj.json \
```
Comment (JZ-LIANG) [inline examples/hybrid_parallelism/pretrain.sh]: ernie base?

**Review by fuyinno4 — APPROVED** [review]

**Review by JZ-LIANG — APPROVED** [review]

**Pull request merged**